File: asciivision/asciivision-wasm/c/ascii.c
#include "ascii.h"

const char ASCII_RAMP_DETAILED[] = " .'`^\",:;Il!i><~+_-?][}{1)(|\\/tfjrxnuvczXYUJCLQ0OZmwqpdbkhao*#MW&8%B@$";
const int ASCII_RAMP_DETAILED_LEN = 70;

const char ASCII_RAMP_SIMPLE[] = " .:-=+*#%@";
const int ASCII_RAMP_SIMPLE_LEN = 10;

char gray_to_ascii(uint8_t gray, int use_detailed) {
    const char* ramp = use_detailed ? ASCII_RAMP_DETAILED : ASCII_RAMP_SIMPLE;
    int ramp_len = use_detailed ? ASCII_RAMP_DETAILED_LEN : ASCII_RAMP_SIMPLE_LEN;
    int index = (gray * (ramp_len - 1)) / 255;
    return ramp[index];
}

File: asciivision/asciivision-wasm/rust/src/ascii.rs
/// Detailed ASCII ramp (70 characters) - from dark to light
pub const ASCII_RAMP_DETAILED: &[u8] =
    b" .'`^\",:;Il!i><~+_-?][}{1)(|\\/tfjrxnuvczXYUJCLQ0OZmwqpdbkhao*#MW&8%B@$";

/// Simple ASCII ramp (10 characters)
pub const ASCII_RAMP_SIMPLE: &[u8] = b" .:-=+*#%@";

/// Map a grayscale value (0-255) to an ASCII character
#[inline]
pub fn gray_to_ascii(gray: u8, use_detailed: bool) -> char {
    let ramp = if use_detailed {
        ASCII_RAMP_DETAILED
    } else {
        ASCII_RAMP_SIMPLE
    };
    let ramp_len = ramp.len();
    let index = (gray as usize * (ramp_len - 1)) / 255;
    ramp[index] as char
}
